### PR TITLE
Revert #474

### PR DESCRIPTION
**Pull request by Gargron**
Reverts tootsuite/documentation#474

I'm sorry but providing a good-to-go nginx configuration is **absolutely necessary**. People can adjust their SSL stuff if they want, but this is a sufficient default. Also the PR seemed kind of mean-spirited.

### CODE REVIEW

**Comment by Dragnucs** [comment]
Can't the doc mention the tool at least?

**Comment by wxcafe** [comment]
I'll concede that the PR was kind of mean spirited. I think if we give people a specific SSL conf we're going to get more attention regarding that TLS part than the rest of the doc, and that discussion is outside of the scope of this repo. We also run the risk of lagging behind and exposing users to security vulns, considering how  the documentation is frequently not completely up-to-date. Either way, it's a problem waiting to happen.

On the other hand, mentionning the tool makes it so that users *have* to check and get some configuration from the mozilla tool (since, by default, it won't run without HTTPS). Admins have to do some manual work to configure mastodon either way, and having them copy-paste from the mozilla tool instead of from this repo isn't that much harder for them, but relieves us of the work and responsibility of maintaining the TLS stuff. It might also make users pay more attention to what they're copying (since they have to go to another place and copy smaller chunks of text at a time), which is a *good thing*.

**Comment by Technowix** [comment]
@wxcafe we can fix this by updating the recommended configuration, and help admins understand what they do with comments like @Aldarone made, but removing stuff will mostly end to people seeking for a thing to copy-paste :v 
And the current configuration is automagically upgrading itself to the best ciphers suits OpenSSL provide.

**Comment by wxcafe** [comment]
There was a huge discussion a while ago about how OpenSSL High wasn't good enough and we should specify select curves for EC ciphers and stuff. 

The advantage of the current state is that people *can't* just blindly copy paste. Anywhere else I'd agree that it's good to have a config that users can just use verbatim but it's plain that it's not possible here (users have to generate certs with Let's Encrypt or something either way, it's never gonna be something they can just copy paste.

This way they can mostly copy paste what the mozilla tool gives them, but they have to go look at it before they just copy the config and I'd hope that would make them look at it a bit.

**Comment by nightpool** [comment]
This PR wasn't merged with the an appropriate amount of time to determine consensus and get feedback from actual admins. The given Mozilla configuration break some android browsers that compromise an outsized portion of the mastodon user base

**Comment by nightpool** [comment]
I think that changing the SSL configuration defaults is fine, and maybe the merged PR is the right way forward, but we should have given a longer time to get feedback from admins first.

**Comment by wxcafe** [comment]
The given mozilla configuration breaks nothing 
screenshot from 2017-12-05 20-29-46

**Comment by Angristan** [comment]
_Re-posting [my message](https://github.com/tootsuite/documentation/pull/474#issuecomment-349437994) here._

Backward support isn't about the cipher here, it's about the TLS version.

Put `EECDH+AESGCM:EECDH+CHACHA20:EECDH+AES` on https://cryptcheck.fr/suite and you will see the cipher is compatible with Android 4.0, IE7, iOS 5, Windows XP... While still being safe. I can't see how Mastodon would work in anything older than that...

There is no hint AES will be broken soon and if it is we're fucked in any case. As for CHACHA20, it' still new, and if you use it on an OpenSSL 1.0 system, which does not support it, it's simply ignored without causing any error.

Drawback of this cipher: none. It's not like a new algorithm will be added tomorrow, and even if it, would have to wait years before seeing it on Debian.

So https://github.com/tootsuite/documentation/commit/72ed10378f4c93eecc901f7e947d7f0508f6345a is useless in termes of useful compatibility and it adds nothing but flaws to the HTTPS config (DSS, DH...).

Concerning [`ssl_ecdh_curve`](https://nginx.org/en/docs/http/ngx_http_ssl_module.html#ssl_ecdh_curve), we should not specify it (as I did before https://github.com/tootsuite/documentation/commit/72ed10378f4c93eecc901f7e947d7f0508f6345a above, so I approve this change), but probably not for the reason you think.

Before the commit, we were using P-256, which does not have any compat issue (even though the author says it has with Android 7, that's not true which makes it look like he really tested his config...).
Why we shouldn't specify one ? Before Nginx 1.11, the default one was `P-256`, so no problem, and since, the default value is `auto`, which equals to  `X25519:P-256:P-521:P-384` with OpenSSL 1.1 and `P-256:P-521:P-384` with older versions. So no compatibility nor security problems here, nothing we nor the users have to worry about. 

And concerning TLS, we have 2 possibility for the doc : 

```
# Add TLSv1.1 and TLSv1.0 for Android 4.0 compatibility 
ssl_protocols TLSv1.2; 
```

or 

```
# Remove TLSv1.0 and TLSv1.1 if you don't need Android 4.0 compatibility
ssl_protocols TLSv1.2 TLSv1.1 TLSv1.0;
```

**TL;DR : Secure, simple, compatible and pretty future-proof config :** 

```
# Add TLSv1.1 and TLSv1.0 for Android 4.0 compatibility 
ssl_protocols TLSv1.2; 
ssl_ciphers EECDH+AESGCM:EECDH+CHACHA20:EECDH+AES;
```

[How it looks on SSL Labs](https://www.ssllabs.com/ssltest/analyze.html?d=mstdn.io&s=164.132.78.232) (see the compatibility section)

And that's all. Compatible with every version of Nginx, and OpenSSL 1.x, so bascially everything. And if the admin wants to use Android 4 phones, he just had to read one line and modify another. I can't see what would be better than this ?



I'm not playing the cryptonerd here, this would be better and easier for everyone, and should not require any modification for years.

What do you think?

*btw, of course I'm not an expert on the subject but there are no personal preferences in this message, just facts. Even so, don't hesitate to correct me.*

**Comment by staticsafe** [comment]
@Angristan there are certain specific versions of Android 7 (7.1.x something), that have issues with P-256.

I know this because a user on tenforward.social specifically mentioned that they were unable to connect to my instance from their 7.1.x device.

Currently I use the Mozilla intermediate profile for my Mastodon instances and on my personal webserver with no complaints from anyone.

If we were to provide a ssl_protocols and ssl_ciphers suggestion, I recommend using that one.

I am not opposed to using the distro provided defaults and letting the admin choose the ssl_protocols and ciphers for themselves.

**Comment by Angristan** [comment]
@staticsafe are we talking about the secp256r1 curve for ECDH ?

If that's the case, you are probably confused. The only issue with ECDH is with Android 7.0 and secp384r1. It was an issue for older version of Nginx because we could only specify one curve, so now we can just use `auto` or add `P-256` at the end to add 7.0 compatibility.

Also, [you are using](https://www.ssllabs.com/ssltest/analyze.html?d=tenforward.social&s=96.126.113.89&latest) P-256 right now (even though DH is still enabled, which it shoud probably not be). P-256 is the shortest curve and the default one. I have never heard about any issue between Android 7 (or any other version) and P-256, but if you have links/proof/whatever to provide, that would be great. 👍

**Comment by staticsafe** [comment]
Ah @Angristan you are correct, I couldn't remember the details of the exact curve so I assumed it was just P-256 in general.

I stand corrected on that issue.
